### PR TITLE
Remove PROTO-serialization from IndexMetaData.Custom

### PR DESCRIPTION
**Pull request by imotov**
Switches IndexMetaData.Custom to the standard named serialization.

Supersedes #32683

@jasontedor this is a version of #32683 without any breaking changes in terms of functionality and user interface. It will require small changes in existing plugins that use IndexMetaData.Custom, but otherwise, I think there is total parity with existing implementation minus PROTO-serialization.

@tlrx could you take a look at my (mis)use of SPI and XContent parsers in [`CreateIndexRequest`](https://github.com/elastic/elasticsearch/compare/master...imotov:remove-proto-from-indexmetadata-custom-2?expand=1#diff-9d81c9f84ef61a9dcbad958e66afb2f3R408) and `PutIndexTemplateRequest`? Did I miss anything there?

### CODE REVIEW

**Comment by imotov** [comment]
retest this please

File: server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java [hunk 1]
```diff
@@ -389,15 +402,20 @@ public CreateIndexRequest source(Map<String, ?> source, DeprecationHandler depre
                 aliases((Map<String, Object>) entry.getValue());
             } else {
                 // maybe custom?
-                IndexMetaData.Custom proto = IndexMetaData.lookupPrototype(name);
-                if (proto != null) {
+                if (entry.getValue() instanceof Map) {
+                    if (customsRegistry == null) {
+                        // Custom registry wasn't specified - we are in client mode let's try loading it from SPI
+                        customsRegistry = new NamedXContentRegistry(getProvidedNamedXContents());
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java]: I'm not sure to understand what you want to achieve here (though I understand the proto based serialization removal).

I don't think we want a request object to create a `NamedXContentRegistry` with only some provided names xcontents as it won't contain all named xcontent parsers but only the ones that provide and implement `NamedXContentProvider`. 

The NamedXContentRegistry is usually instantiated once on the server or client side and it must contain all required named xcontent parsers. This list is potentially enriched with named xcontent parsers provided by modules or plugins that implement `NamedXContentProvider`. But in any case, the NamedXContentRegistry is instantiated once on server or client side with everything it needs.

On server side the registry is instantiated in the [Node class](https://github.com/elastic/elasticsearch/blob/db6e8c736d92582fe56024993450ce0a987b498d/server/src/main/java/org/elasticsearch/node/Node.java#L394-L401) and in the [RestHighLevelClient class](https://github.com/elastic/elasticsearch/blob/db6e8c736d92582fe56024993450ce0a987b498d/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java#L236-L238) for an example of client side instanciation.

In your case, I guess you'll have to plumb the `NamedXContentRegistry` down to where you want to parse this content. I think it's also possible to retrieve a ref on the current registry from a XContentParser instance.

Comment (imotov) [reply server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java]: Thanks! As per our discussion, I have removed the automatic `NamedXContentRegistry` discovery and replaced with passing it to `source(...)` method as parameters.

**Comment by imotov** [comment]
retest this please

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/NamedXContentRegistry.java [hunk 2]
```diff
@@ -141,4 +141,15 @@ public NamedXContentRegistry(List<Entry> entries) {
         return categoryClass.cast(entry.parser.parse(parser, context));
     }
 
+    /**
+     * Returns true if the given named object is supported by the current stream
+     */
+    public <T> boolean supportsNamedObject(Class<T> categoryClass, String name) {
+        Map<String, Entry> parsers = registry.get(categoryClass);
+        if (parsers == null || registry.isEmpty()) {
```
Comment (dakrone) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/NamedXContentRegistry.java]: I think the `registry.isEmpty()` check is superfluous here

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 3]
```diff
@@ -80,57 +88,40 @@
...
-    }
-
-    public static <T extends Custom> T lookupPrototypeSafe(String type) {
-        //noinspection unchecked
-        T proto = (T) customPrototypes.get(type);
-        if (proto == null) {
-            throw new IllegalArgumentException("No custom metadata prototype registered for type [" + type + "]");
+    public static Custom parseCustom(String name, Map<String, Object> map, DeprecationHandler deprecationHandler,
+                                     NamedXContentRegistry customsRegistry) throws IOException {
+        XContentBuilder customXContent = XContentFactory.jsonBuilder()
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Can you put this in a try-with-resources block?

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java [hunk 4]
```diff
@@ -469,13 +467,13 @@ public static IndexTemplateMetaData fromXContent(XContentParser parser, String t
...
                     } else {
                         // check if its a custom index metadata
-                        IndexMetaData.Custom proto = IndexMetaData.lookupPrototype(currentFieldName);
-                        if (proto == null) {
-                            //TODO warn
+                        try {
+                            IndexMetaData.Custom custom = parser.namedObject(IndexMetaData.Custom.class, currentFieldName, null);
+                            builder.putCustom(custom.getWriteableName(), custom);
+                        } catch (UnknownNamedObjectException ex) {
+                            logger.warn("Skipping unknown custom index metadata object with type {}", currentFieldName);
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java]: I think this should be lowercase (suuuuper minor)

**Review by dakrone — APPROVED** [review]
This LGTM, thanks Igor, I left a couple tiny comments.

@jasontedor might want to take a look and make sure this will be workable for CCR though

**Comment by dakrone** [comment]
@jasontedor will this work for CCR's needs?

**Comment by dakrone** [comment]
Okay I've made some changes to this PR:

This PR nows removes the deprecated `Custom` class in `IndexMetaData`, in favor
of a `Map<String, DiffableStringMap>` that is used to store custom index
metadata. As part of this, there is now no way to set this metadata in a
template or create index request (since it's only set by plugins, or dedicated
REST endpoints).

The `Map<String, DiffableStringMap>` is intended to be a namespaced `Map<String,
String>` (`DiffableStringMap` implements `Map<String, String>`, so the signature
is more like `Map<String, Map<String, String>>`). This is so we can do things
like:

``` java
Map<String, String> ccrMeta = indexMetaData.getCustom("ccr");
```

And then have complete control over the metadata. This also means any
plugin/feature that uses this has to manage its own BWC, as the map is just
serialized as a map. It also means that if metadata is put in the map that isn't
used (for instance, if a plugin were removed), it causes no failures the way
an unregistered `Setting` would.

The reason I use a custom `DiffableStringMap` here rather than a plain
`Map<String, String>` is so the map can be diffed with previous cluster state
updates for serialization.

/ping @martijnvg @jasontedor @s1monw as I believe you were all interested in the implementation of this

File: server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This is a {@code Map<String, String>} that implements AbstractDiffable so it
+ * can be used for cluster state purposes
+ */
+public class DiffableStringMap extends AbstractDiffable<DiffableStringMap> implements Map<String, String> {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java]: Can we extend from `AbstractMap` here instead of `AbstractDiffable` and then implement `Diffable` instead of `Map`? I think that would simplify this class and the only method that `AbstractDiffable` provides here is overwritten here.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 6]
```diff
@@ -519,13 +464,12 @@ public MappingMetaData mappingOrDefault(String mappingType) {
...
-    public ImmutableOpenMap<String, Custom> getCustoms() {
-        return this.customs;
+    public ImmutableOpenMap<String, DiffableStringMap> getCustomData() {
+        return this.customData;
     }
 
-    @SuppressWarnings("unchecked")
-    public <T extends Custom> T custom(String type) {
-        return (T) customs.get(type);
+    public Map<String, String> getCustomData(final String key) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Are we returning here a mutable map? I don't think we should do that here. We should at least wrap it in `Collections#unmodifiableMap()`

File: server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        upserts.put("newkey", "yay");
+        assertThat(dsmd.getUpserts(), equalTo(upserts));
+
+        DiffableStringMap dsm3 = diff.apply(dsm);
+        logger.info("--> 3: {}", dsm3);
+        assertThat(dsm3.get("foo"), equalTo("not-bar"));
+        assertThat(dsm3.get("newkey"), equalTo("yay"));
+        assertThat(dsm3.get("baz"), equalTo("eggplant"));
+        assertThat(dsm3.get("potato"), equalTo(null));
+    }
```
Comment (martijnvg) [inline server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java]: Can we also add a test here with a few more maps and entries (maybe randomly generated) than in the existing test?

Comment (martijnvg) [reply server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java]: and maybe also a test that verifies that the serialization works?

**Review by martijnvg — COMMENTED** [review]
@dakrone This looks great! I left a couple of comments.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 6 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Yep good idea, I'll do that

File: server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java (hunk 5 above)
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java]: +1

File: server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java [hunk 8]
```diff
@@ -474,11 +448,10 @@ public void readFrom(StreamInput in) throws IOException {
...
         }
-        int customSize = in.readVInt();
-        for (int i = 0; i < customSize; i++) {
-            String type = in.readString();
-            IndexMetaData.Custom customIndexMetaData = IndexMetaData.lookupPrototypeSafe(type).readFrom(in);
-            customs.put(type, customIndexMetaData);
+        if (in.getVersion().before(Version.V_7_0_0_alpha1)) {
+            // This used to be the size of custom metadata classes
+            int customSize = in.readVInt();
+            assert customSize == 0 : "unexpected custom metadata when none is supported";
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java]: hmm do we need to skip the size if we are in production? I mean that assert will not trip if we run without -ea

File: server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java (hunk 7 above)
Comment (s1monw) [reply server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java]: I also wonder if we need a random test that adds randomly to one map and then passes it on as a diff to another and makes sure after we did that N times the maps are the same?

File: server/src/test/resources/META-INF/services/org.elasticsearch.plugins.spi.NamedXContentProvider [hunk 9]
```diff
@@ -1 +1 @@
-org.elasticsearch.plugins.spi.NamedXContentProviderTests$TestNamedXContentProvider
\ No newline at end of file
+org.elasticsearch.plugins.spi.NamedXContentProviderTests$TestNamedXContentProvider
```
Comment (s1monw) [inline server/src/test/resources/META-INF/services/org.elasticsearch.plugins.spi.NamedXContentProvider]: this is an unrelated change?

File: server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        m.put("baz", "eggplant");
+        m.put("potato", "canon");
+        DiffableStringMap dsm = new DiffableStringMap(m);
+
+        Map<String, String> m2 = new HashMap<>();
+        m2.put("foo", "not-bar");
+        m2.put("newkey", "yay");
+        m2.put("baz", "eggplant");
+        DiffableStringMap dsm2 = new DiffableStringMap(m2);
+        logger.info("--> 1: {}", dsm);
```
Comment (s1monw) [inline server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java]: can we remove the logging in this unittest?

**Review by s1monw — APPROVED** [review]
Looks good in general. I left some minors and once @martijnvg comments are resolved I am good to go with that. nice and simple!

File: server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java (hunk 5 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java]: Yep that definitely works, will do that

File: server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java (hunk 7 above)
Comment (dakrone) [reply server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java]: > and maybe also a test that verifies that the serialization works?

The serialization was added as part of the `IndexMetaDataTests`, but I'll add one here for targeted serialization

File: server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java (hunk 8 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java]: I'm not following, do you want it to be a regular exception instead of an assert? We still need to do the `readVInt` regardless of `-ea` or without, it's just the 0 size comparison that is an assert for tests

File: server/src/test/resources/META-INF/services/org.elasticsearch.plugins.spi.NamedXContentProvider (hunk 9 above)
Comment (dakrone) [reply server/src/test/resources/META-INF/services/org.elasticsearch.plugins.spi.NamedXContentProvider]: Yep sorry, this was part of reverting Igor's changes that missed a newline

File: server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java (hunk 10 above)
Comment (dakrone) [reply server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java]: Sure, will do this

File: server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java (hunk 7 above)
Comment (dakrone) [reply server/src/test/java/org/elasticsearch/cluster/metadata/DiffableStringMapTests.java]: > I also wonder if we need a random test that adds randomly to one map and then passes it on as a diff to another and makes sure after we did that N times the maps are the same?

I'll add a test for that as well

File: server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java (hunk 8 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java]: I think that we need to do a hard exception indeed, we won't be able to read the rest of the message at this point because we can't read the customs (no `IndexMetaData#lookupPrototypeSafe anymore`) when there are unexpected ones. That is, `customSize > 0` is fatal for us.

File: server/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java [hunk 11]
```diff
@@ -494,11 +474,10 @@ public void readFrom(StreamInput in) throws IOException {
...
         }
-        int customSize = in.readVInt();
-        for (int i = 0; i < customSize; i++) {
-            String type = in.readString();
-            IndexMetaData.Custom customIndexMetaData = IndexMetaData.lookupPrototypeSafe(type).readFrom(in);
-            customs.put(type, customIndexMetaData);
+        if (in.getVersion().before(Version.V_7_0_0_alpha1)) {
+            // Used to be used for custom index metadata
+            int customSize = in.readVInt();
+            assert customSize == 0 : "expected not to have any custom metadata";
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java]: This should be fatal.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 12]
```diff
@@ -780,10 +714,14 @@ public static IndexMetaData readFrom(StreamInput in) throws IOException {
...
-            Custom customIndexMetaData = lookupPrototypeSafe(type).readFrom(in);
-            builder.putCustom(type, customIndexMetaData);
+        if (in.getVersion().onOrAfter(Version.V_7_0_0_alpha1)) {
+            for (int i = 0; i < customSize; i++) {
+                String key = in.readString();
+                DiffableStringMap custom = new DiffableStringMap(in);
+                builder.putCustom(key, custom);
+            }
+        } else {
+            assert customSize == 0 : "expected no custom index metadata";
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: This should be fatal.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java [hunk 13]
```diff
@@ -227,11 +210,10 @@ public static IndexTemplateMetaData readFrom(StreamInput in) throws IOException
...
         }
-        int customSize = in.readVInt();
-        for (int i = 0; i < customSize; i++) {
-            String type = in.readString();
-            IndexMetaData.Custom customIndexMetaData = IndexMetaData.lookupPrototypeSafe(type).readFrom(in);
-            builder.putCustom(type, customIndexMetaData);
+        if (in.getVersion().before(Version.V_7_0_0_alpha1)) {
+            // Previously we allowed custom metadata
+            int customSize = in.readVInt();
+            assert customSize == 0 : "expected no custom metadata";
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java]: This should be fatal.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left one comment.

**Comment by dakrone** [comment]
Thanks @jasontedor, I've addressed your comments

**Review by jasontedor — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+    public Diff<DiffableStringMap> diff(DiffableStringMap previousState) {
+        return new DiffableStringMapDiff(previousState, this);
+    }
+
+    public static Diff<DiffableStringMap> readDiffFrom(StreamInput in) throws IOException {
+        return new DiffableStringMapDiff(in);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java]: I don't think it is necessary to overwrite equals(), hashCode() and toString() here? `AbstractMap` already implements these methods.

**Comment by martijnvg** [comment]
LGTM - left one minor comment.
